### PR TITLE
feat: add per-dataset timestep_sampling_offset for semantic-aware flow-matching training

### PR DESCRIPTION
**Pull request by okdsf**
## Summary

Add a `timestep_sampling_offset` option to per-dataset config (`multidatabackend.json`). This offsets the pre-sigmoid normal sample in the default logit-normal flow-matching schedule, letting users shift the noise distribution per dataset based on semantic granularity.

- Negative offset → lower-noise timesteps (detail-focused)
- Positive offset → higher-noise timesteps (structure-focused)
- Default `0.0` = stock behavior

Based on *The Quadratic Geometry of Flow Matching: Semantic Granularity Alignment for Text-to-Image Synthesis* ([arXiv:2603.10785](https://arxiv.org/abs/2603.10785)).

## Motivation

Even at the same resolution, images with different semantic content benefit from different noise emphasis. Close-up/fine-detail images benefit from lower-noise sampling (detail refinement), while full-body/macro-structure images benefit from higher-noise sampling (structural learning). For FLUX on anime data, per-content noise offset also stabilizes training by smoothing overly aggressive gradient updates from the strong photoreal prior.

## Usage

```json
[
  {
    "id": "closeup_data",
    "type": "local",
    "instance_data_dir": "/data/closeups",
    "timestep_sampling_offset": -0.5
  },
  {
    "id": "fullbody_data",
    "type": "local",
    "instance_data_dir": "/data/fullbody",
    "timestep_sampling_offset": 0.5
  }
]
```

## Results

Anime LoRA training on FLUX.1 dev, ~1400 images, 3 characters. Left = baseline, right = with offset.

<table>
<tr><th>Baseline</th><th>With timestep_sampling_offset</th></tr>
<tr><td></td><td></td></tr>
<tr><td></td><td></td></tr>
<tr><td></td><td></td></tr>
</table>

Improved facial detail, sharper hands/fingers, more precise accessories, better lighting.

## Changes

- `simpletuner/helpers/models/common.py` — `_get_dataset_timestep_bias()` helper + pre-sigmoid offset in the default sigmoid sampling branch of `sample_flow_sigmas`
- `simpletuner/helpers/data_backend/factory.py` — parse `timestep_sampling_offset` from config (mirrors `probability`)
- `simpletuner/helpers/data_backend/config/image.py` — dataclass field + `from_dict` + `to_dict`
- `simpletuner/simpletuner_sdk/server/data/dataset_blueprints.py` — WebUI field + default
- `documentation/DATALOADER.md` + zh/ja/es/pt-BR/hi translations — document the new option
- `tests/test_backend_config.py` — config parse/roundtrip/default assertions
- `tests/test_timestep_bias_sampling.py` — unit tests for `_get_dataset_timestep_bias`

## Testing

- `test_backend_config`: 48 tests passed (including new `timestep_sampling_offset` assertions)
- `test_timestep_bias_sampling`: 3 tests passed (reads config, defaults to 0.0, handles missing backend_id)
- GPU verification: real `sample_flow_sigmas` on RTX 5090, 16384 samples, bias direction correct, unset == stock (bit-identical)
- Smoke train: Anima Base v1.0 LoRA, 4 steps, `timestep_sampling_offset=-0.5`, `SMOKE_EXIT=0`

Fully opt-in, default `0.0` = zero behavioral change. Only the default sigmoid schedule is affected; uniform/beta/custom-timestep schedules are unchanged.

### CODE REVIEW

**Comment by bghira** [comment]
you should try stacking T-LoRA on top of this for rank-aware masking by timestep as well, if you're into schedule modifications that improve composition vs coherence.

**Comment by okdsf** [comment]
Thanks for the pointer! I looked into T-LoRA (Soboleva et al., AAAI 2026) — very interesting work.

Our motivation for timestep sampling offset actually addresses a related problem. For existing models — whether community-mature anime models (SDXL-based) or photoreal models like FLUX — the pretrained prior is very strong. During fine-tuning, the model tends to get trapped in this prior, and generated images often carry a heavy imprint of the pretrained style. We introduce per-dataset timestep offset to smooth the learning trajectory by controlling how often each noise level is sampled for different semantic content, helping the model move past the prior toward the target distribution more gracefully.

T-LoRA's core finding — that overfitting primarily occurs at higher (noisier) timesteps — corroborates our observations. In anime fine-tuning scenarios though, the behavior is more nuanced:

- **FLUX**: the strong photoreal prior causes the model to memorize local details (faces, textures) first and defer structural learning, often resulting in stiff body poses. This aligns with T-LoRA's observation that coarse-structure high-noise steps are overfitting-prone, though the root cause likely extends beyond rank capacity to gradient conflicts between the photoreal prior and anime distribution.
- **SDXL (U-Net)**: possibly due to local receptive field bias, anime fine-tuning tends toward color oversaturation and cluttered compositions — a different failure mode from what DiT's global-attention architecture exhibits.

The two approaches are complementary: T-LoRA controls *how many parameters update per step* (rank masking), while sampling offset controls *how often each noise level is sampled*. Combining them would be an interesting direction to explore.

**Comment by bghira** [comment]
the one thing i'm confused about, your DATALOADER.md update says timestep_bias and the PR description says timestep_sampling_offset

**Comment by bghira** [comment]
if you trained your test models using `timestep_sampling_offset`, i don't think anything happened - extraneous values are simply ignored in dataloader configs. you may be seeing training seed level variance if that's the case, you should also set `seed` to a non-null value like `42` to have reproducible training runs.

**Comment by okdsf** [comment]
Good catch — you're right, the code and docs still use the old name `timestep_bias` while the PR description says `timestep_sampling_offset`. I submitted this feature in parallel to kohya sd-scripts (kohya-ss/sd-scripts#2401), where a community reviewer suggested renaming from `timestep_bias` to `timestep_sampling_offset` to align with the existing `noise_offset` naming convention. I updated the PR description here but forgot to sync the code and documentation.

Regarding the smoke train validity: the code parses `timestep_bias` and the dataloader config also uses `timestep_bias`, so the feature did take effect. I'll now unify everything to `timestep_sampling_offset` and re-verify.

**Comment by okdsf** [comment]
42 — great number! Seems like everyone uses it 😄

Here are results under seed=42, FLUX.1 dev LoRA training on anime data.

FLUX has a very strong photoreal prior, which means the convergence basin for anime style is extremely narrow.

**Q2 (step 800), Q3 (step 900)** — both with `timestep_sampling_offset` enabled. Even with just 100 steps apart, the quality difference is significant — this demonstrates how sharp FLUX's peak effect is. The model is extremely sensitive near the optimal region, even with offset enabled.

<table>
<tr><th>Q2 (step 800, with offset)</th><th>Q3 (step 900, with offset)</th></tr>
<tr><td></td><td></td></tr>
</table>

**Q1** — left: with offset, right: without offset. The right side has a noticeably more painterly/oil-painting look — the model failed to escape the photoreal prior's convergence basin. The left side (with offset) successfully converged to the anime style.



The core purpose of this feature is to **smooth the training trajectory**, helping the model transition past the pretrained prior's convergence basin toward the target style more gracefully.

Sigma distribution histogram showing the offset effect (N=50000, logit-normal schedule):
